lib/remoteproc/remoteproc_virtio.c

```diff
@@ -16,6 +16,92 @@
 #include <metal/utilities.h>
 #include <metal/alloc.h>
 
+static void rproc_virtio_delete_virtqueues(struct virtio_device *vdev)
+{
+	struct virtio_vring_info *vring_info;
+	unsigned int i;
+
+	if (vdev->vrings_info) {
+		for (i = 0; i < vdev->vrings_num; i++) {
+			vring_info = &vdev->vrings_info[i];
+			if (vring_info->vq) {
+				virtqueue_free(vring_info->vq);
+			}
+		}
+	}
+}
+
+static int rproc_virtio_create_virtqueue(struct virtio_device *vdev,
+					 unsigned int flags,
+					 unsigned int idx,
+					 const char *name,
+					 vq_callback callback)
+{
+	struct virtio_vring_info *vring_info;
+	struct vring_alloc_info *vring_alloc;
+	int ret;
+	(void)flags;
+
+	/* Get the vring information */
+	vring_info = &vdev->vrings_info[idx];
+	vring_alloc = &vring_info->info;
+
+	/* Alloc the virtqueue and init it */
+	metal_assert(!vring_info->vq);
+	vring_info->vq = virtqueue_allocate(vring_alloc->num_descs);
+	if (!vring_info->vq) {
+		return ERROR_NO_MEM;
+	}
+
+#ifndef VIRTIO_DEVICE_ONLY
+	if (vdev->role == VIRTIO_DEV_DRIVER) {
+		size_t offset = metal_io_virt_to_offset(vring_info->io,
+							vring_alloc->vaddr);
+		size_t size = vring_size(vring_alloc->num_descs,
+					 vring_alloc->align);
+		metal_io_block_set(vring_info->io, offset, 0, size);
+	}
+#endif
+	ret = virtqueue_create(vdev, idx, name, vring_alloc, callback,
+			       vdev->func->notify, vring_info->vq);
+	if (ret) {
+		return ret;
+	}
```

Comment:
useless brackets + 1 blank line before the return

Reply:
Done